subsys/caf/modules/sensor_data_aggregator.c

```diff
@@ -0,0 +1,209 @@
+	[i].buf_count = DT_INST_PROP(i, buf_count),					\
+	[i].buf_len = DT_INST_PROP(i, buf_data_length),					\
+	[i].agg_buffers = (struct aggregator_buffer *) &agg_ ## i ## _bufs,		\
+	[i].active_buf = (struct aggregator_buffer *) &agg_ ## i ## _bufs,
+
+#define __DEFINE_AGG							\
+	static struct aggregator aggregators[] = {			\
+	DT_INST_FOREACH_STATUS_OKAY(__DEFINE_AGGREGATOR)		\
+	}
+
+struct aggregator_buffer {
+	uint8_t *data;		/* Dynamic data. */
+	bool busy;		/* Buffer status. */
+	uint8_t sample_cnt;	/* Number of samples already saved in the buffer. */
+};
+
+struct aggregator {
+	const char *sensor_descr;			/* sensor_description of the sensor. */
+	struct aggregator_buffer *agg_buffers;	/* Buffers */
+	struct aggregator_buffer *active_buf;	/* Active buffer to which data will be placed */
+	enum sensor_state sensor_state;		/* Sensors state */
+	const uint8_t sensor_data_size;		/* Size of sensor data in bytes. */
+	const uint8_t buf_count;		/* Number of buffers */
+	const uint8_t buf_len;			/* Size of buffor data in bytes. */
+};
+
+
+DT_INST_FOREACH_STATUS_OKAY(__DEFINE_BUF_DATA) //no semicolon on purpose.
+__DEFINE_AGG;
+
+
+static struct aggregator_buffer *get_free_buffer(struct aggregator *agg)
+{
+	for (size_t i = 0; i < agg->buf_count; i++) {
+		if (agg->agg_buffers[i].busy == false) {
+			return &agg->agg_buffers[i];
+		}
+	}
+	return NULL;
+}
+
+static struct aggregator *get_aggregator(const char *sensor_descr)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(aggregators); i++) {
+		if (sensor_descr == aggregators[i].sensor_descr) {
+			return &aggregators[i];
+		}
+	}
+	return NULL;
+}
+
+static void clean_agg_buffer(struct aggregator *agg, struct aggregator_buffer *ab)
+{
+	__ASSERT_NO_MSG(ab);
+
+	LOG_INF("Cleaning, sample_cnt %d", ab->sample_cnt);
+	ab->sample_cnt = 0;
+	ab->busy = false;
+	if (agg->active_buf == NULL) {
+		agg->active_buf = ab;
+	}
+}
+
+static void send_agg_buff_data(struct aggregator *agg, struct aggregator_buffer *ab)
+{
+	struct sensor_aggregator_event *event = new_sensor_aggregator_event();
+
+	event->buf = ab->data;
+	event->sample_cnt = ab->sample_cnt;
+	event->sensor_status = agg->sensor_state;
+	event->sensor_descr = agg->sensor_descr;
+	EVENT_SUBMIT(event);
+}
+
+static int add_sample_to_aggregator(struct sensor_event *event, struct aggregator *agg)
+{
+	if (event->dyndata.size != agg->sensor_data_size) {
+		return -EBADMSG;
+	}
+	if (!agg->active_buf) {
+		return -ENOMEM;
+	}
+	struct aggregator_buffer *ab = agg->active_buf;
+
+	size_t pos = ab->sample_cnt * agg->sensor_data_size;
+	size_t avail = agg->buf_len - pos;
+
+	if (avail < agg->sensor_data_size) {
+		__ASSERT_NO_MSG(false);
+		return -ENOMEM;
+	}
+
+	memcpy(&ab->data[pos], event->dyndata.data, event->dyndata.size);
+	ab->sample_cnt++;
+	avail -= event->dyndata.size;
+
+	if (avail < agg->sensor_data_size) {
+		ab->busy = true;
+		agg->active_buf = get_free_buffer(agg);
+		send_agg_buff_data(agg, ab);
```

Comment:
I would swap lines 142 and 143 - it will logically divide space on what is done on old buffer and where we select new one